File: web/lib/photo-url.ts
export function safePhotoUrl(source: string): string | null {
  try {
    const url = new URL(source);
    if (url.protocol !== 'https:' || url.hostname !== 'res.cloudinary.com' || url.port || url.username || url.password || !url.pathname.startsWith('/dskoyv2oe/image/upload/')) return null;
    return url.toString();
  } catch { return null; }
}
/** Preserve stored originals; request raster previews for formats not shared by all clients. */
export function photoPreviewUrl(source: string): string {
  if (!safePhotoUrl(source)) return '';
  try {
    const url = new URL(source);
    if (
      url.protocol !== 'https:' ||
      url.hostname !== 'res.cloudinary.com' ||
      !url.pathname.startsWith('/dskoyv2oe/image/upload/')
    )
      return source;
    if (!/\.(svg|heic|heif|tif|tiff|avif)$/i.test(url.pathname)) return source;
    if (url.pathname.includes('/f_png/')) return source;
    url.pathname = url.pathname.replace(
      '/image/upload/',
      '/image/upload/f_png/',
    );
    return url.toString();
  } catch {
    return source;
  }
}

File: app/src/main/java/com/example/onthejob/data/upload/PhotoUrl.kt
package com.example.onthejob.data.upload

import java.net.URI

/** Use a compatible raster preview without changing the stored original URL. */
fun photoPreviewUrl(source: String): String {
    return try {
        val uri = URI(source)
        val path = uri.rawPath ?: return source
        if (uri.scheme != "https" || uri.host != "res.cloudinary.com" ||
            !path.startsWith("/dskoyv2oe/image/upload/") ||
            !Regex("\\.(svg|heic|heif|tif|tiff|avif)$", RegexOption.IGNORE_CASE).containsMatchIn(path) ||
            path.contains("/f_png/")) return source
        source.replaceFirst("/image/upload/", "/image/upload/f_png/")
    } catch (_: Exception) { source }
}
